router/http.go

```diff
@@ -275,62 +303,28 @@ func (s *HTTPListener) findRouteForHost(host string) *httpRoute {
 	return nil
 }
 
-func fail(sc *httputil.ServerConn, req *http.Request, code int, msg string) {
-	resp := &http.Response{
-		StatusCode:    code,
-		ProtoMajor:    1,
-		ProtoMinor:    0,
-		Request:       req,
-		Body:          ioutil.NopCloser(bytes.NewBufferString(msg)),
-		ContentLength: int64(len(msg)),
-	}
-	sc.Write(req, resp)
+func fail(w http.ResponseWriter, code int, msg string) {
+	w.Header().Set("Content-Length", strconv.Itoa(len(msg)))
+	w.WriteHeader(code)
+	w.Write([]byte(msg))
 }
 
-var errMissingTLS = errors.New("router: route not found or TLS not configured")
-
-func (s *HTTPListener) handle(conn net.Conn, isTLS bool) {
-	defer conn.Close() // TODO(benburkert): for TLS, check that closing this conn sends the TLS "close notify"
-
-	var r *httpRoute
+const hdrUseStickySessions = "Flynn-Use-Sticky-Sessions"
 
-	if isTLS {
-		certForHandshake := func(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
-			r = s.findRouteForHost(hello.ServerName)
-			if r == nil {
-				return nil, errMissingTLS
-			}
-			return r.keypair, nil
-		}
-		conn = tls.Server(conn, tlsconfig.SecureCiphers(&tls.Config{
-			GetCertificate: certForHandshake,
-			Certificates:   []tls.Certificate{s.keypair},
-		}))
+func (s *HTTPListener) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	r := s.findRouteForHost(req.Host)
+	if r == nil {
+		fail(w, 404, "Not Found")
+		return
 	}
 
-	sc := httputil.NewServerConn(conn, nil)
-	for {
-		req, err := sc.Read()
-		if err != nil {
-			if err != io.EOF && err != httputil.ErrPersistEOF && err != errMissingTLS {
-				log.Println("client read err:", err)
-			}
-			return
-		}
-
-		if !isTLS {
-			r = s.findRouteForHost(req.Host)
-			if r == nil {
-				fail(sc, req, 404, "Not Found")
-				continue
-			}
-		}
-
-		req.RemoteAddr = conn.RemoteAddr().String()
-		if r.service.handle(req, sc, isTLS, r.Sticky) {
-			return
-		}
+	// TODO(bgentry): find a better way to access this setting in the service
+	// where it's needed.
+	if r.Sticky {
+		req.Header.Set(hdrUseStickySessions, "true")
```

Comment:
+1 for using a context here. In the mean time, always set this header so that it can't be injected.